### DISCUSSION TITLE
Populate details Grid Based on selected lookup value in master dialog

### DISCUSSION THREAD

**Discussion by rhemasystems** [serenity-is/Serenity#6664] [Help] [answered]
I have a Lookup dropdown  in Dialog.ts What I want to do is I want to populate Grid in the same Dialog.ts based on the value selected in drop down.
I am aware of the **Filtered Lookup in Detail Dialog** in the sample demo but that doesnt fully address my issue.
If there is no data available based on the selected value, the grid should be blank, and only populate if there is corresponding details data available in the database.
So once the drop down is selected I want to load data from the database based on the record selected, otherwise the grid should be empty.

![image](https://user-images.githubusercontent.com/50826372/213880878-00abce2a-536b-438e-96ee-ce3792da1a63.png)

**Comment by rhemasystems** [comment, marked as answer]
I will really appreciate any kind of help. Thank you

**Reply by arunkumar12712** [reply]
let purchageItems: purchageItems[] = [];
purchageItems  push value from database  
this.form.LineItems.value = purchageItems;

**Reply by rhemasystems** [reply]
Thank you arunkumar12712, could you please explain some more, how do I use the service to select only the columns that I need to populate the LineItems grid

**Reply by muhammedyaman** [reply]
you can use lookup to achieve this too. 
```
...
 this.form.PurchaseField.changeSelecet2(e=>{
    var purchaseItems = PurchaseItemsRow.getLookup().filter(q=>q.PurchaseId == this.form.PurchaseField.value);
            for (var item in purchaseItems ) {
                this.form.IcerikList.view.addItem({
                    firstField: item.firstField,
                    ...
                });
            }
});
```

or you can use setItems() function in detailgrid. But you must first get your purchaseitems and modify that into an array of DetailRow.

**Reply by rhemasystems** [reply]
Thanks but after implelemnting I get this error
`Message: Uncaught Each data element must implement a unique '__id' property. Object at index '1' has no identity value: {"Id":1,"ProductName":"A4 Sheets","Quantity":1}`

My Code:
` var purchaseItems = SmartERP.Purchases.PurchaseDetailsRow.getLookup().items.filter(q => q.PurchaseId == purchaseId);`
` purchaseItems.forEach(element => {`
    `   this.form.LineItems.view.addItem({`
       `     Id:element.Id,`
        `     ProductName: element.ProductName,`
        `     Quantity: element.Quantity`
    `   });`
`});`

**Reply by muhammedyaman** [reply]
You have to set new Id for the detail rows. 
I am using these functions from [this beautiful extra](https://github.com/dfaruque/Serenity.Extra) which are: 
```
public nextId = 1;
        public getNextId() {
            return "`" + this.nextId++;
        }
        public setNewId(entity: TEntity) {
            entity[this.getIdProperty()] = this.getNextId();
        }
```
if you define these in your editor you will be able to use it like this: 
```
purchaseItems.forEach(element => {
    var lineItem = this.form.LineItems.setNewId({
     Id:element.Id,
     ProductName: element.ProductName,
     Quantity: element.Quantity
   })
   this.form.LineItems.view.addItem(lineItem);
```

**Reply by arunkumar12712** [reply]
try this
purchaseservice.retrieve({
entityid: set your dropdown values
}, res=>{
      res.entity.purchaseItem.forach((element)=>{
         lineItem.push({
           set value all fields
         })
      })
}, { asys : false})

**Reply by muhammedyaman** [reply]
> try this purchaseservice.retrieve({ entityid: set your dropdown values }, res=>{ res.entity.purchaseItem.forach((element)=>{ lineItem.push({ set value all fields }) }) }, { asys : false})

I dont think this is correct since he is trying to insert new rows to the detail field. But if you want to use service you need to use the List method and send needed parameters such as purchaseId.

**Reply by rhemasystems** [reply]
> try this purchaseservice.retrieve({ entityid: set your dropdown values }, res=>{ res.entity.purchaseItem.forach((element)=>{ lineItem.push({ set value all fields }) }) }, { asys : false})

I am having an issue with this line `res.entity.purchaseItem.forach((element)=>{` there is the red error line under `purchaseItem`
am a bit lost which method to try now
Please help

**Reply by muhammedyaman** [reply]
In your LineItemsEditor.ts define this function so you can reach setNewId function from your Master dialog: 
```
public publicSetNewId(item: LineItemsRow) {
            super.setNewId(item);
        }
```
in your MasterDialog.ts constructor or afterloadentity:
```
this.form.Purchase.changeSelect2(e => {
                XYZService.List({ Criteria: [Serenity.Criteria("PurchaseId"), '=', [this.form.Purchase.value]] }, resp => {
                    for (var entity in resp.Entities) {
                        var item: LineItemsRow = {                            
                            ProductName: entity.ProductName,
                            Quantity: entity.Quantity
                        }
                        this.form.LineItems.publicSetNewId(item);
                        this.form.LineItems.view.addItem(item);
                    }
                })
            })
```
XYZService is the service where you have the data you want after you select from purchase lookup. 
As I mentioned before you can use lookup instead of Service but I wanted to give you an example of Service usage.

**Reply by rhemasystems** [reply]
Thank you  muhammdyaman, it worked perfectly, but one question, how do I clear the grid before inserting into it so  that in case the user changes the value it should clear the grid before adding the new records

**Reply by muhammedyaman** [reply]
You're most welcome! 
Just add this line: 
`this.form.LineItems.setItems([]);`
before this one:
`XYZService.List({ Criteria: [Serenity.Criteria("PurchaseId"), '=', [this.form.Purchase.value]] }, resp => {...`

**Reply by rhemasystems** [reply]
Thank you so much muhammedyaman, it worked.

**Reply by muhammedyaman** [reply]
My pleasure.
